### PR TITLE
support manifestlist images

### PR DESCRIPTION
**Pull request by liulangwa**
support build windows platform

### CODE REVIEW

**Comment by liulangwa** [comment]
限制了只能同步amd64和armd64的arch，这个修改可以编译windows运行的程序

**Review by hhyasdf — COMMENTED** [review]
What an excellent job, Thanks for your contribution!

I have some feedbacks and we can have discussion for all of them.

File: pkg/sync/manifest.go [hunk 1]
```diff
@@ -1,55 +1,54 @@
...
 		}
 
-		manifestDigests := []*digest.Digest{}
-		for _, item := range ml.Manifests {
-			d, err := digest.Parse(item.Digest)
-			if err != nil {
-				return nil, nil, err
+		for _, manifestDescriptorElem := range manifestSchemaListInfo.Manifests {
+
+			// 暂时至支持arm64/amd64架构的镜像同步
```
Comment (hhyasdf) [inline pkg/sync/manifest.go:37]: Could you please use English comments here? That's more friendly for all of people. :)

File: pkg/sync/manifest.go [hunk 2]
```diff
@@ -1,55 +1,54 @@
...
+			if manifestDescriptorElem.Platform.Architecture == "amd64" || manifestDescriptorElem.Platform.Architecture == "arm64" {
+				// 用指定Arch的Digest再拉一次针对该Digest的多层Manifest
```
Comment (hhyasdf) [inline pkg/sync/manifest.go:39]: English comments here plz. :)

File: pkg/sync/source.go [hunk 3]
```diff
@@ -94,30 +94,24 @@ func (i *ImageSource) GetBlobInfos(manifestByte []byte, manifestType string) ([]
...
-	}
-
-	// get a manifest
-
-	blobInfos := manifestInfo.LayerInfos()
+	// get a Blobs
 	srcBlobs := []types.BlobInfo{}
-
+	for _, manifestInfo := range manifestInfoSlice {
+	blobInfos := manifestInfo.LayerInfos()
```
Comment (hhyasdf) [inline pkg/sync/source.go:105]: Code indentation should be adjusted for L105~103. :)

File: pkg/sync/task.go [hunk 4]
```diff
@@ -72,18 +76,85 @@ func (t *Task) Run() error {
...
+		if err != nil {
+			return err
+		}
+
+		var subManifestByte []byte
+
+		// push manifest to destination
+		for _, manifestDescriptorElem := range manifestSchemaListInfo.Manifests {
+
+			if manifestDescriptorElem.Platform.Architecture == "amd64" || manifestDescriptorElem.Platform.Architecture == "arm64" {
```
Comment (hhyasdf) [inline pkg/sync/task.go:94]: Will we get into trouble if not only for "amd64" and "arm64" ? I'm wondering if other architectures can also be supported here.

**Review by hhyasdf — COMMENTED** [review]
Besides the changes, I think is better to leave the vendor in this repository for compiling easily, you can use something like "go mod vendor" to generate.

**Comment by hhyasdf** [comment]
> 限制了只能同步amd64和armd64的arch，这个修改可以编译windows运行的程序

What's the exactly problem for windows? Maybe we can look into it together.

File: pkg/sync/task.go (hunk 4 above)
Comment (liulangwa) [reply pkg/sync/task.go:94]: no trouble to support other arch, just for save storage space

**Comment by liulangwa** [comment]
> > 限制了只能同步amd64和armd64的arch，这个修改可以编译windows运行的程序
> 
> What's the exactly problem for windows? Maybe we can look into it together.

it because some old lib not support windows, so

**Comment by liulangwa** [comment]
> Besides the changes, I think is better to leave the vendor in this repository for compiling easily, you can use something like "go mod vendor" to generate.

The vendor is removed in order not to submit  code of lib that has nothing to do with the function of the program

**Comment by hhyasdf** [comment]
> > Besides the changes, I think is better to leave the vendor in this repository for compiling easily, you can use something like "go mod vendor" to generate.
> 
> The vendor is removed in order not to submit code of other library that has nothing to do with the function of the program

I think it might be better to use `go mod tidy & go mod vendor` to remove useless dependents, rather than remove the whole vendor.

File: pkg/sync/task.go (hunk 4 above)
Comment (hhyasdf) [reply pkg/sync/task.go:94]: image-syncer use system memory to sync images, maybe it's not necessary.

**Comment by hhyasdf** [comment]
> > > 限制了只能同步amd64和armd64的arch，这个修改可以编译windows运行的程序
> > 
> > 
> > What's the exactly problem for windows? Maybe we can look into it together.
> 
> it because some old lib not support windows, so

I think most of our user may use Linux for development rather than Windows, maybe we can put window compilation support to further plan.

File: pkg/sync/task.go (hunk 4 above)
Comment (liulangwa) [reply pkg/sync/task.go]: i have removed the limit

**Comment by liulangwa** [comment]
> > > Besides the changes, I think is better to leave the vendor in this repository for compiling easily, you can use something like "go mod vendor" to generate.
> > 
> > 
> > The vendor is removed in order not to submit code of other library that has nothing to do with the function of the program
> 
> I think it might be better to use `go mod tidy & go mod vendor` to remove useless dependents, rather than remove the whole vendor.

ok

**Comment by hhyasdf** [comment]
Well done! Thanks for your contribution!

**Pull request merged**